### PR TITLE
rec: Add bindings to set arbitrary key-value metadata in logged messages

### PR DESCRIPTION
**Pull request by chbruyand**
### Short description

This PR adds new protobuf `meta` field that can contain arbitrary key-value metadata.

This still needs:
- [x] regression tests
- [ ] the corresponding non FFI interface ?

New protobuf `meta` field can also be read as the following protov3 map:
```
  message MetaValue {
    repeated string stringVal = 1;
    repeated int64 intVal = 2;
  }

  map<string, MetaValue> meta = 22;
```  
### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)
- [ ]  checked that this code was merged to master

### CODE REVIEW

File: .not-formatted [hunk 1]
```diff
@@ -42,47 +42,19 @@
 ./pdns/devpollmplexer.cc
```
Comment (omoerbeek) [inline .not-formatted:42]: Hmm, sort -c agrees here on the old order, using the C locale. How did you sort this file? It's hard to see what's the actual changes are right now.

**Review by omoerbeek — COMMENTED** [review]
Rest of review will follow once this is non-Draft.

File: .not-formatted (hunk 1 above)
Comment (chbruyand) [reply .not-formatted:42]: Good catch! Wasn't using the C locale.

File: pdns/lua-recursor4.cc [hunk 2]
```diff
@@ -251,148 +263,133 @@ void RecursorLua4::postPrepareContext()
...
-    {"NSIP",       (int)DNSFilterEngine::PolicyType::NSIP       }
-    }});
-
-  for(const auto& n : QType::names)
+    });
+
+  d_lw->registerFunction("check", (bool (SuffixMatchNode::*)(const DNSName&) const) & SuffixMatchNode::check);
+  d_lw->registerFunction("toString", (string(SuffixMatchNode::*)() const) & SuffixMatchNode::toString);
+
+  d_pd.push_back({"policykinds", in_t{{"NoAction", (int)DNSFilterEngine::PolicyKind::NoAction}, {"Drop", (int)DNSFilterEngine::PolicyKind::Drop}, {"NXDOMAIN", (int)DNSFilterEngine::PolicyKind::NXDOMAIN}, {"NODATA", (int)DNSFilterEngine::PolicyKind::NODATA}, {"Truncate", (int)DNSFilterEngine::PolicyKind::Truncate}, {"Custom", (int)DNSFilterEngine::PolicyKind::Custom}}});
```
Comment (omoerbeek) [inline pdns/lua-recursor4.cc:364]: I don ;t like the long line. Maybe mark this with
```
// clang-format off
...
// clang-format on
```

File: pdns/lua-recursor4.cc [hunk 3]
```diff
@@ -251,148 +263,133 @@ void RecursorLua4::postPrepareContext()
...
+
+  d_pd.push_back({"policytypes", in_t{{"None", (int)DNSFilterEngine::PolicyType::None}, {"QName", (int)DNSFilterEngine::PolicyType::QName}, {"ClientIP", (int)DNSFilterEngine::PolicyType::ClientIP}, {"ResponseIP", (int)DNSFilterEngine::PolicyType::ResponseIP}, {"NSDName", (int)DNSFilterEngine::PolicyType::NSDName}, {"NSIP", (int)DNSFilterEngine::PolicyType::NSIP}}});
```
Comment (omoerbeek) [inline pdns/lua-recursor4.cc:366]: Same here

File: pdns/pdns_recursor.cc [hunk 4]
```diff
@@ -1018,6 +1019,11 @@ static void protobufLogQuery(LocalStateHolder<LuaConfigItems>& luaconfsLocal, co
   if (!policyTags.empty()) {
     m.addPolicyTags(policyTags);
   }
+  if (!meta.empty()) {
```
Comment (omoerbeek) [inline pdns/pdns_recursor.cc:1022]: If is redundant

**Review by omoerbeek — COMMENTED** [review]
Looks good to me. A few nits and docs are missing. Also, it would be good if somebody  more knowledgeable about ffi interfaces takes a look.

File: pdns/lua-recursor4.cc [hunk 5]
```diff
@@ -175,23 +186,23 @@ void RecursorLua4::postPrepareContext()
...
   d_lw->registerMember("data", &DNSQuestion::data);
-  d_lw->registerMember<uint16_t (DNSQuestion::*)>("extendedErrorCode", [](const DNSQuestion& dq) -> uint16_t {
+  d_lw->registerMember<uint16_t(DNSQuestion::*)>(
+    "extendedErrorCode", [](const DNSQuestion& dq) -> uint16_t {
       if (dq.extendedErrorCode && *dq.extendedErrorCode) {
         return *(*dq.extendedErrorCode);
       }
-      return 0;
-    },
+      return 0; },
```
Comment (rgacogne) [inline pdns/lua-recursor4.cc:194]: Wow, did our formatting tool do that? That's awful.

Comment (rgacogne) [reply pdns/lua-recursor4.cc:194]: Well in fact most of the formatting changes in that file are terrible.

File: pdns/pdns_recursor.cc [hunk 6]
```diff
@@ -1721,6 +1730,7 @@ static void startDoResolve(void *p)
     dq.proxyProtocolValues = &dc->d_proxyProtocolValues;
     dq.extendedErrorCode = &dc->d_extendedErrorCode;
     dq.extendedErrorExtra = &dc->d_extendedErrorExtra;
+    dq.meta = dc->d_meta;
```
Comment (rgacogne) [inline pdns/pdns_recursor.cc:1733]:
```suggestion
    dq.meta = std::move(dc->d_meta);
```

**Review by rgacogne — APPROVED** [review]
Looks good, I wonder if we should exclude `lua-recursor4.cc` from the formatting, at least partially.

**Review by rgacogne — APPROVED** [review]

**Comment by chbruyand** [comment]
I turned off the clang formating for the whole `RecursorLua4::postPrepareContext()` function and kept it on for the rest of the file. This looks indeed better like that.

File: pdns/recursordist/docs/lua-scripting/ffi.rst [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+   Disable padding
+
+.. function:: pdns_ffi_param_add_record(pdns_ffi_param_t* ref, const char* name, uint16_t type, uint32_t ttl, const char* content, size_t contentSize, pdns_record_place_t place) -> bool
+
+   Adds a record. Returns true if it was correctly added, false otherwise
+
+.. function:: pdns_ffi_param_add_meta_single_string_kv(pdns_ffi_param_t* ref, const char* key, const char* val) -> void
+
+    .. versionadded:: 4.5.3
```
Comment (omoerbeek) [inline pdns/recursordist/docs/lua-scripting/ffi.rst:135]: Does this need a backport? If so, the earliest version it's going to be in is 4.5.5, as 4.5.4 was just released without this. Otherweise the `versionadded::`  should be 4.6.0

File: pdns/recursordist/docs/lua-scripting/ffi.rst [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+   This function allows you to add an arbitrary string value for a given key in the ``meta`` field of the produced :doc:`protobuf <../lua-config/protobuf>` log message
+
+.. function:: pdns_ffi_param_add_meta_single_int64_kv(pdns_ffi_param_t *ref, const char* key, int64_t val) -> void
+
+    .. versionadded:: 4.5.3
```
Comment (omoerbeek) [inline pdns/recursordist/docs/lua-scripting/ffi.rst:141]: same here

**Review by omoerbeek — APPROVED** [review]
LGTM!

**Pull request merged**